### PR TITLE
Add tzlocal singleton

### PR DESCRIPTION
**Pull request by alimcmaster1**
Use a metaclass to implement the singleton pattern for tzlocal class.

Partially closes <#668>

### CODE REVIEW

File: dateutil/tz/_factories.py [hunk 1]
```diff
@@ -11,6 +13,29 @@ def __call__(cls):
...
+class _TzChangeSingleton(type):
+
+    """
+    A :class: `tzlocalsingleton`
+               Singleton that guards against TZ changes
+    """
+
+    def __init__(cls, *args, **kwargs):
+        cls.__instance = None
+        cls.__tz = time.tzname
```
Comment (pganssle) [inline dateutil/tz/_factories.py]: This guard is not as robust as one would hope, though it's probably a pretty good sentinel, consider:

```python
with TZEnvContext('EST5EDT'):
    l1 = tz.tzlocal()

with TZEnvContext('EST6EDT'):
    l2 = tz.tzlocal()
```

**Comment by alimcmaster1** [comment]
Thanks @pganssle for the review. Updated as per comments and added test case. What do you think?

**Comment by pganssle** [comment]
This is clever, but I'm not sure how fast it is.

That said, it's also got a problem, because setting the `TZ` variable doesn't instantly change the results of `time.tzname` and the other things that `tzlocal` looks for, it only changes after `time.tzset` has been called, so this will fail if you did this:

```python
import os
from dateutil import tz

os.environ['TZ'] = 'PST8PDT'
local0 = tz.tzlocal()
os.environ['TZ'] = 'EST5EDT'
local1 = tz.tzlocal()
time.tzset()
local2 = tz.tzlocal()
```

In that situation, `local0`, `local1` and `local2` will all be PST8PDT, because the cache will be invalidated between `local0` and `local1`, but the actual underlying hooks will not be returning anything different, and `local2` will hit the (incorrect) cache.

**Comment by pganssle** [comment]
Whether or not it is ambiguous (I'm not sure I've ever seen it actually cause a problem), populating the cache based on changes in `os.environ` would severely exacerbate the problem, because normally:

```python
os.environ['TZ'] = 'EST5EDT'
local1 = tz.tzlocal()
time.tzset()
local2 = tz.tzlocal()
```

In the above situation, at the very least `local2` would be correct. With this cache, all subsequent `tzlocal` calls would return the broken version unless you change `TZ` to a third thing and back.

**Comment by alimcmaster1** [comment]
Thanks okay understood definitely agree.

I can’t see a solution here apart from perhaps calling time.tzset() from within the cache if we detect a change in ‘TZ’ environment var since the last call to the cache. 

But this would be a breaking change if anyone has any functionality like the snippet below.

Think I’ll leave this here for now unless anyone else has any ideas?

**Comment by alimcmaster1** [comment]
Spoke offline about this closing PR

**Pull request closed without merging**